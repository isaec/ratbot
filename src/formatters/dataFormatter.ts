import {
  Divider,
  Section,
  section,
  md,
  divider,
  url,
} from "@src/helpers/slack/blocks";
import {
  HyperlinkedPurchaseRequestData,
  PurchaseRequestKeys,
} from "../gsheetReader";
import { formatRange } from "./stringFormatter";

const read = (val: string | undefined) => val ?? "N/A";

const displayedKeys: PurchaseRequestKeys[] = ["Price Per", "Qty", "Comments"];

export type SlackBlock = Array<Divider | Section>;

export const formatData = ({
  data,
  hyperlink,
}: HyperlinkedPurchaseRequestData): SlackBlock => [
  section([
    md(`*${read(data["Line #"])}:* ${url(read(data["Item"]), hyperlink)}`),
    ...displayedKeys
      .filter((key) => data[key] !== undefined)
      .map((key) => md(`*${key}:* ${data[key]}`)),
  ]),
];

const defined = <T>(x: T | undefined): x is T => x !== undefined;

const transformer =
  <T, K>(fn: (arg0: T) => K) =>
  (arg0: T) =>
    fn(arg0);
const toInt = transformer(parseInt);

const moneyFormatter = new Intl.NumberFormat("en-US", {
  style: "currency",
  currency: "USD",
});

export const formatDataArray = (
  dataArray: HyperlinkedPurchaseRequestData[]
): SlackBlock => [
  ...dataArray.flatMap((data) => [...formatData(data), divider()]),
  section(
    md(
      `*Aggregate of request for ${formatRange(
        // may become expensive compared to single iteration
        dataArray
          .map((obj) => obj.data["Line #"])
          .filter(defined)
          .map(toInt)
      )}*`
    ),
    [
      md(
        `*Total Price + Tax:* ${moneyFormatter.format(
          dataArray
            .map((obj) => obj.data["Price + Tax"])
            .filter(defined)
            .map((price) => parseFloat(price.replace("$", "")))
            .reduce((a, b) => a + b, 0)
        )}`
      ),
    ]
  ),
];

// hacky function
export const formattedDataArrayToString = (
  data: ReturnType<typeof formatDataArray>
) => {
  let str = "";
  for (const block of data) {
    if (block.type === "section") {
      if (block.text !== undefined && typeof block.text.text === "string") {
        str += block.text.text + "\n";
      }
      if (block.fields) {
        for (const field of block.fields) {
          str += field.text + "\n";
        }
      }
    } else if (block.type === "divider") {
      str += "----\n";
    }
  }
  return str.trim();
};
