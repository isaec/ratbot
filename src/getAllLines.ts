/**
 * matches "line 1234" but not "line 1234 and"
 */
const singleLine = /lines?\s(\d+)(?:\s(?!-|\d|and|&)|$)/g;

const singleHyphenatedLine = /lines?\s(\d+)-(?!\d)/g;

const lineRange = /lines?\s(\d+)-(\d+)/g;

const lineAndLine = /lines?\s(\d+)\s(?:and|&|\+)\s(\d+)/g;

const runRegexes = (
  string: string,
  results: Set<number>,
  ...regexes: RegExp[]
) => {
  for (const regex of regexes) {
    for (const match of string.matchAll(regex)) {
      if (typeof match[1] === "string") {
        results.add(parseInt(match[1]));
      }
    }
  }
};

export const getAllLines = (text: string): Set<number> => {
  const lines: Set<number> = new Set();
  // run the regex that return a single line match
  runRegexes(text, lines, singleLine, singleHyphenatedLine);
  for (const match of text.matchAll(lineRange)) {
    if (typeof match[1] === "string" && typeof match[2] === "string") {
      const start = parseInt(match[1]);
      const end = parseInt(match[2]);
      for (let i = start; i <= end; i++) lines.add(i);
    }
  }
  for (const match of text.matchAll(lineAndLine)) {
    if (typeof match[1] === "string" && typeof match[2] === "string") {
      lines.add(parseInt(match[1]));
      lines.add(parseInt(match[2]));
    }
  }

  return lines;
};

export const accumulate = <T>(iter: Iterator<T>): T[] => {
  const result: T[] = [];
  let next = iter.next();
  while (!next.done) {
    result.push(next.value);
    next = iter.next();
  }
  return result;
};

export const setToSortedArray = <T>(set: Set<T>): T[] =>
  accumulate(set.values()).sort();
