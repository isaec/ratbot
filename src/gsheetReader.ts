import * as dotenv from "dotenv";
import { GoogleSpreadsheet, GoogleSpreadsheetRow } from "google-spreadsheet";
dotenv.config();

const doc = new GoogleSpreadsheet(process.env.GSHEET_SHEET_ID);

const assertDefined = <T>(value: T | undefined): T => {
  if (value === undefined) {
    throw new Error("Value is undefined");
  }
  return value;
};

console.log(
  "serviceAccount.client_email",
  process.env.GSHEET_SERVICE_ACCOUNT_CLIENT_EMAIL
);

await doc.useServiceAccountAuth({
  client_email: assertDefined(process.env.GSHEET_SERVICE_ACCOUNT_CLIENT_EMAIL),
  private_key: atob(
    assertDefined(process.env.GSHEET_SERVICE_ACCOUNT_PRIVATE_KEY_BTOA)
  ),
});
await doc.loadInfo();
const sheet = await doc.sheetsByTitle[
  process.env.GSHEET_SHEET_TITLE ?? "purchase requests"
];

const headerRowIndex = parseInt(process.env.GSHEET_HEADER_ROW ?? "1");
const urlColumnIndex = parseInt(process.env.GSHEET_URL_COLUMN ?? "9");

await sheet.loadHeaderRow(headerRowIndex);

// /^\s+(?:'|\s)([^':]*).*/gm
export type PurchaseRequestKeys =
  | "Line #"
  | "Status"
  | "Request Date"
  | "Approval Date"
  | "Ordered Date"
  | "Requestor Name"
  | "Item"
  | "Supplier"
  | "URL"
  | "Qty"
  | "Price Per"
  | "Total Price (no tax)"
  | "Total Price Paid (including shipping)"
  | "Price + Tax"
  | "PO?"
  | "Invoice Sent"
  | "Lead"
  | "Lead Mentor"
  | "Category"
  | "Comments";

export type PurchaseRequestData = Partial<Record<PurchaseRequestKeys, string>>;

type PurchaseRequestRow = GoogleSpreadsheetRow & PurchaseRequestData;

const computeOffset = (rowNumber: number) => {
  return rowNumber - (headerRowIndex + 1);
};

const getRow = async (rowNumber: number) => {
  const rows = await sheet.getRows({
    offset: computeOffset(rowNumber),
    limit: 1,
  });
  return rows[0] as PurchaseRequestRow;
};

const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const cachedCellSet = new Set<number>();

const getHyperlink = async (rowNumber: number) => {
  if (!cachedCellSet.has(rowNumber)) {
    const position = `${alphabet.charAt(urlColumnIndex)}${rowNumber}`;
    console.log("loading", { position });
    await sheet.loadCells(position);
    cachedCellSet.add(rowNumber);
  }

  return sheet.getCell(rowNumber - 1, urlColumnIndex).hyperlink;
};

export type HyperlinkedPurchaseRequestData = {
  data: PurchaseRequestData;
  hyperlink: string | undefined;
};

export const readSheetLine = async (
  line: number
): Promise<HyperlinkedPurchaseRequestData> => {
  const row = await getRow(line);
  const hyperlink = await getHyperlink(line);
  console.log({ line, hyperlink });
  return {
    data: row,
    hyperlink,
  };
};
