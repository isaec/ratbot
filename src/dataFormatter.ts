import { PurchaseRequestData, PurchaseRequestKeys } from "./gsheetReader";

const r = (val: string | undefined) => val ?? "N/A";

const displayedKeys: PurchaseRequestKeys[] = [
  "Requestor Name",
  "Supplier",
  "Price Per",
  "Qty",
  "Total Price (no tax)",
  "Price + Tax",
  "Category",
  "Comments",
];

export const formatData = (data: PurchaseRequestData) => [
  {
    type: "section",
    text: {
      type: "mrkdwn",
      text: `*${r(data["Line #"])}:* ${r(data["Item"])}`,
    },
    fields: displayedKeys.map((key) => ({
      type: "mrkdwn",
      text: `*${key}:* ${r(data[key])}`,
    })),
  },
  {
    type: "section",
    text: {
      type: "mrkdwn",
      text: `*URL:* ${r(data["URL"])}`,
    },
  },
];

export const formatDataArray = (dataArray: PurchaseRequestData[]) =>
  dataArray.flatMap((data) => [...formatData(data), { type: "divider" }]);

export const formattedDataArrayToString = (
  data: ReturnType<typeof formatDataArray>
) => {
  let str = "";
  for (const block of data) {
    if (block.type === "section") {
      if (block.text !== undefined && typeof block.text.text === "string") {
        str += block.text.text + "\n";
      }
      if (block.fields) {
        for (const field of block.fields) {
          str += field.text + "\n";
        }
      }
    } else if (block.type === "divider") {
      str += "----\n";
    }
  }
  return str;
};
