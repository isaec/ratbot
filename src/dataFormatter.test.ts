import { assert, describe, expect, test } from "vitest";
import { formatData, formattedDataArrayToString } from "./dataFormatter";
import { getAllLines, setToSortedArray } from "./getAllLines";
import { generate, optional } from "generate-combinations";
import { PurchaseRequestData } from "./gsheetReader";

describe("formatData", () => {
  test.each(
    generate<PurchaseRequestData>({
      "Requestor Name": optional("ian"),
      Supplier: optional("andy mark"),
      "Price Per": optional("$9.99"),
      Qty: optional("1"),
      "Total Price (no tax)": optional("9.99"),
      "Price + Tax": optional("10.73"),
      Category: optional("FRC Robots"),
      Comments: optional("this is a comment"),
    })
  )("formatData(%s)", (data: PurchaseRequestData) => {
    expect(formatData(data)).toMatchSnapshot();
  });
});

describe("formattedDataArrayToString", () => {
  test.each([
    `[{"type":"section","text":{"type":"mrkdwn","text":"*35:* 6 Gauge Connector 1/4 Stud Hole"},"fields":[{"type":"mrkdwn","text":"*Requestor Name:* Daniel"},{"type":"mrkdwn","text":"*Supplier:* andymark"},{"type":"mrkdwn","text":"*Price Per:* $3.10"},{"type":"mrkdwn","text":"*Qty:* 20"},{"type":"mrkdwn","text":"*Total Price (no tax):* $62.00"},{"type":"mrkdwn","text":"*Price + Tax:* $67.97"},{"type":"mrkdwn","text":"*Category:* Extra Materials"},{"type":"mrkdwn","text":"*Comments:* N/A"}]},{"type":"section","text":{"type":"mrkdwn","text":"*URL:* https://www.andymark.com/products/compression-lug-connector-6-gauge-1-4-stud-hole-burndy-yazv6ctc14fx?via=Z2lkOi8vYW5keW1hcmsvV29ya2FyZWE6OkNhdGFsb2c6OkNhdGVnb3J5LzViYjYxODZhYmM2ZjZkNmRlMWU2OWY5Yg"}},{"type":"divider"}]`,
  ])(`formattedDataArrayToString(%s)`, (objStr) => {
    expect(formattedDataArrayToString(JSON.parse(objStr))).toMatchSnapshot();
  });
});
