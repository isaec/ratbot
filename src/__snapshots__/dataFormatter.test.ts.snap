// Vitest Snapshot v1

exports[`formattedDataArrayToString([{"type":"section","text":{"type":"mrkdwn","text":"*35:* 6 Gauge Connector 1/4 Stud Hole"},"fields":[{"type":"mrkdwn","text":"*Requestor Name:* Daniel"},{"type":"mrkdwn","text":"*Supplier:* andymark"},{"type":"mrkdwn","text":"*Price Per:* $3.10"},{"type":"mrkdwn","text":"*Qty:* 20"},{"type":"mrkdwn","text":"*Total Price (no tax):* $62.00"},{"type":"mrkdwn","text":"*Price + Tax:* $67.97"},{"type":"mrkdwn","text":"*Category:* Extra Materials"},{"type":"mrkdwn","text":"*Comments:* N/A"}]},{"type":"section","text":{"type":"mrkdwn","text":"*URL:* https://www.andymark.com/products/compression-lug-connector-6-gauge-1-4-stud-hole-burndy-yazv6ctc14fx?via=Z2lkOi8vYW5keW1hcmsvV29ya2FyZWE6OkNhdGFsb2c6OkNhdGVnb3J5LzViYjYxODZhYmM2ZjZkNmRlMWU2OWY5Yg"}},{"type":"divider"}]) 1`] = `
"*35:* 6 Gauge Connector 1/4 Stud Hole
*Requestor Name:* Daniel
*Supplier:* andymark
*Price Per:* $3.10
*Qty:* 20
*Total Price (no tax):* $62.00
*Price + Tax:* $67.97
*Category:* Extra Materials
*Comments:* N/A
*URL:* https://www.andymark.com/products/compression-lug-connector-6-gauge-1-4-stud-hole-burndy-yazv6ctc14fx?via=Z2lkOi8vYW5keW1hcmsvV29ya2FyZWE6OkNhdGFsb2c6OkNhdGVnb3J5LzViYjYxODZhYmM2ZjZkNmRlMWU2OWY5Yg
----
"
`;
